ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java

```diff
@@ -459,7 +461,11 @@ private void readImpl(DataStreamRequestByteBuf request, ChannelHandlerContext ct
       localWrite = CompletableFuture.completedFuture(0L);
       remoteWrites = Collections.emptyList();
     } else if (request.getType() == Type.STREAM_DATA) {
-      localWrite = info.getLocal().write(request.slice(), request.getWriteOptionList(), writeExecutor);
+      if (close && request.getDataLength() == 0) {
+        localWrite = CompletableFuture.completedFuture(0L);
+      } else {
+        localWrite = info.getLocal().write(request.slice(), request.getWriteOptionList(), writeExecutor);
+      }
```

Comment:
This part is an optimization for not calling `getLocal().write(..)`.  However, `getLocal().write(..)` also takes care the `WriteOption`s such as `SYNC` and `FLUSH`.  So, we cannot skip it here.  We could skip it later on in
```java
//writeTo(..)
@@ -301,6 +307,9 @@ public class DataStreamManagement {
     final DataChannel channel = stream.getDataChannel();
     long byteWritten = 0;
     for (ByteBuffer buffer : buf.nioBuffers()) {
+      if (buffer.remaining() == 0) {
+        continue;
+      }
       final ReferenceCountedObject<ByteBuffer> wrapped = ReferenceCountedObject.wrap(
           buffer, buf::retain, ignored -> buf.release());
       try(UncheckedAutoCloseable ignore = wrapped.retainAndReleaseOnClose()) {
```

Reply:
And the length check here, maybe we can remove it? since it will always get into here. https://github.com/apache/ratis/blob/master/ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java#L322

Reply:
For length check, do you mean the new code?
```java
if (buffer.remaining() == 0) {
   continue;
}
```
It is just a minor improvement.  The value `buffer.remaining()` is likely non-zero for most cases.  I am fine if we don't want to add it.